src/fu-main.c

```diff
@@ -403,21 +412,233 @@ fu_main_authorize_install_cb (GObject *source, GAsyncResult *res, gpointer user_
-				&error)) {
-		g_dbus_method_invocation_return_gerror (helper->invocation, error);
+	/* do the next authentication action ID */
+	fu_main_authorize_install_queue (g_steal_pointer (&helper));
+}
+
+static void
+fu_main_authorize_install_queue (FuMainAuthHelper *helper_ref)
+{
+	FuMainPrivate *priv = helper_ref->priv;
+	g_autoptr(FuMainAuthHelper) helper = helper_ref;
+	g_autoptr(GError) error = NULL;
+
+	/* still more things to to authenticate */
+	if (helper->action_ids->len > 0) {
+		g_autofree gchar *action_id = g_strdup (g_ptr_array_index (helper->action_ids, 0));
+		g_autoptr(PolkitSubject) subject = g_object_ref (helper->subject);
+		g_ptr_array_remove_index (helper->action_ids, 0);
+		polkit_authority_check_authorization (priv->authority, subject,
+						      action_id, NULL,
+						      POLKIT_CHECK_AUTHORIZATION_FLAGS_ALLOW_USER_INTERACTION,
+						      NULL,
+						      fu_main_authorize_install_cb,
+						      g_steal_pointer (&helper));
 		return;
 	}
 
+	/* all authenticated, so install all the things */
+	for (guint i = 0; i < helper->install_tasks->len; i++) {
+		FuInstallTask *task = g_ptr_array_index (helper->install_tasks, i);
+		if (!fu_engine_install (helper->priv->engine,
+					fu_install_task_get_device (task),
+					fu_install_task_get_app (task),
+					helper->blob_cab,
+					helper->flags,
+					&error)) {
+			g_dbus_method_invocation_return_gerror (helper->invocation, error);
+			return;
+		}
+	}
+
 	/* success */
 	g_dbus_method_invocation_return_value (helper->invocation, NULL);
 }
 
+static const GError *
+fu_main_error_array_find (GPtrArray *errors, FwupdError error_code)
+{
+	for (guint j = 0; j < errors->len; j++) {
+		const GError *error = g_ptr_array_index (errors, j);
+		if (g_error_matches (error, FWUPD_ERROR, error_code))
+			return error;
+	}
+	return NULL;
+}
+
+static guint
+fu_main_error_array_count (GPtrArray *errors, FwupdError error_code)
+{
+	guint cnt = 0;
+	for (guint j = 0; j < errors->len; j++) {
+		const GError *error = g_ptr_array_index (errors, j);
+		if (g_error_matches (error, FWUPD_ERROR, error_code))
+			cnt++;
+	}
+	return cnt;
+}
+
+static gboolean
+fu_main_error_array_matches_any (GPtrArray *errors, FwupdError *error_codes)
+{
+	for (guint j = 0; j < errors->len; j++) {
+		const GError *error = g_ptr_array_index (errors, j);
+		gboolean matches_any = FALSE;
+		for (guint i = 0; error_codes[i] != FWUPD_ERROR_LAST; i++) {
+			if (g_error_matches (error, FWUPD_ERROR, error_codes[i])) {
+				matches_any = TRUE;
+				break;
+			}
+		}
+		if (!matches_any)
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static GError *
+fu_main_error_array_build_composite (GPtrArray *errors)
+{
+	FwupdError err_prio[] =		{ FWUPD_ERROR_INVALID_FILE,
+					  FWUPD_ERROR_VERSION_SAME,
+					  FWUPD_ERROR_VERSION_NEWER,
+					  FWUPD_ERROR_NOT_SUPPORTED,
+					  FWUPD_ERROR_INTERNAL,
+					  FWUPD_ERROR_NOT_FOUND,
+					  FWUPD_ERROR_LAST };
+	FwupdError err_all_uptodate[] =	{ FWUPD_ERROR_VERSION_SAME,
```

Comment:
I'd think you probably also want to match VERSION_NEWER in this case right?

Reply:
I don't think so. "All updatable firmware is already installed" isn't right if you're trying to install `firmware-1.2.3.cab` when you have 1.2.4 installed. Maybe we could add a new `err_all_newer` with  message like "All updatable firmware is already a newer version" or something like that?

Reply:
Yeah that's a good alternative.

Reply:
Fixed.